io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -125,7 +123,10 @@ data TimerCompletionInfo s =
        Timer !(TVar s TimeoutState)
      | TimerRegisterDelay !(TVar s Bool)
      | TimerThreadDelay !ThreadId
+     | TimerTimeout !ThreadId !TimeoutId
```

Comment:
So my suggestion is to put the lock in here, as a `STVar s Bool`.

Reply:
I actually prefer the solution I went for, I think it is simpler and avoids having to use STM for handling timeouts inside IOSim, do you think it's worth to change?